File: src/flow/flow.ts
/**
 * Flow data model — the central unit of captured traffic.
 * Mirrors mitmproxy's HTTPFlow = Request + Response (+ error, websocket, metadata).
 */

import { Headers } from "./headers.ts";
import { randomUUID } from "node:crypto";

export interface ClientInfo {
  address: string;
  port: number;
  tls: boolean;
}

export class CnRequest {
  scheme: "http" | "https" = "http";
  host = "";
  port = 80;
  method = "GET";
  path = "/";
  httpVersion = "1.1";
  headers = new Headers();
  body: Buffer | null = null;
  /** True when the body was dropped because it exceeded maxBodySize. */
  bodyTruncated = false;
  timestampStart = 0;
  timestampEnd = 0;

  get url(): string {
    const defaultPort = this.scheme === "https" ? 443 : 80;
    const hostPart = this.port === defaultPort ? this.host : `${this.host}:${this.port}`;
    return `${this.scheme}://${hostPart}${this.path}`;
  }

  get prettyHost(): string {
    return this.headers.get("host")?.split(":")[0] || this.host;
  }
}

export class CnResponse {
  statusCode = 0;
  reason = "";
  httpVersion = "1.1";
  headers = new Headers();
  body: Buffer | null = null;
  bodyTruncated = false;
  timestampStart = 0;
  timestampEnd = 0;

  get contentType(): string {
    return this.headers.get("content-type") || "";
  }
}

export interface WebSocketMessage {
  fromClient: boolean;
  /** "text" | "binary" */
  type: "text" | "binary";
  content: Buffer;
  timestamp: number;
}

export class FlowError {
  constructor(public msg: string, public timestamp: number) {}
}

export type FlowType = "http" | "websocket";

/** Interception state: how a paused flow should be resumed. */
type ResumeResolver = (action: "resume" | "kill") => void;

export class Flow {
  readonly id: string;
  type: FlowType = "http";
  request: CnRequest;
  response: CnResponse | null = null;
  error: FlowError | null = null;
  client: ClientInfo;
  websocketMessages: WebSocketMessage[] = [];

  // metadata
  timestampCreated: number;
  marked = false;
  comment = "";
  /** Names of rules that touched this flow (for UI badges). */
  appliedRules: string[] = [];
  /** True when this flow's response was synthesized (mocked) rather than fetched. */
  mocked = false;

  // interception
  intercepted = false;
  private resumeResolver: ResumeResolver | null = null;

  constructor(client: ClientInfo, now: number, id?: string) {
    this.id = id ?? randomUUID();
    this.request = new CnRequest();
    this.client = client;
    this.timestampCreated = now;
  }

  get duration(): number | null {
    if (!this.response) return null;
    return this.response.timestampEnd - this.request.timestampStart;
  }

  /** Pause this flow; await the returned promise at a streaming boundary. */
  intercept(): Promise<"resume" | "kill"> {
    this.intercepted = true;
    return new Promise<"resume" | "kill">((resolve) => {
      this.resumeResolver = resolve;
    });
  }

  resume(): void {
    if (this.resumeResolver) {
      this.intercepted = false;
      this.resumeResolver("resume");
      this.resumeResolver = null;
    }
  }

  kill(): void {
    if (this.resumeResolver) {
      this.intercepted = false;
      this.resumeResolver("kill");
      this.resumeResolver = null;
    }
  }

  /** Lossless serialization for persistence (full fields + base64 bodies). */
  toRecord(): FlowRecord {
    const reqHdr = this.request.headers.entries();
    return {
      id: this.id,
      type: this.type,
      timestampCreated: this.timestampCreated,
      marked: this.marked,
      comment: this.comment,
      mocked: this.mocked,
      appliedRules: this.appliedRules,
      error: this.error ? { msg: this.error.msg, timestamp: this.error.timestamp } : null,
      client: this.client,
      request: {
        scheme: this.request.scheme,
        host: this.request.host,
        port: this.request.port,
        method: this.request.method,
        path: this.request.path,
        httpVersion: this.request.httpVersion,
        headers: reqHdr,
        body: this.request.body ? this.request.body.toString("base64") : null,
        bodyTruncated: this.request.bodyTruncated,
        timestampStart: this.request.timestampStart,
        timestampEnd: this.request.timestampEnd,
      },
      response: this.response
        ? {
            statusCode: this.response.statusCode,
            reason: this.response.reason,
            httpVersion: this.response.httpVersion,
            headers: this.response.headers.entries(),
            body: this.response.body ? this.response.body.toString("base64") : null,
            bodyTruncated: this.response.bodyTruncated,
            timestampStart: this.response.timestampStart,
            timestampEnd: this.response.timestampEnd,
          }
        : null,
      websocketMessages: this.websocketMessages.map((m) => ({
        fromClient: m.fromClient,
        type: m.type,
        content: m.content.toString("base64"),
        timestamp: m.timestamp,
      })),
    };
  }

  /** Reconstruct a flow from a persisted record. */
  static fromRecord(rec: FlowRecord): Flow {
    const flow = new Flow(rec.client, rec.timestampCreated, rec.id);
    flow.type = rec.type;
    flow.marked = rec.marked;
    flow.comment = rec.comment;
    flow.mocked = rec.mocked;
    flow.appliedRules = rec.appliedRules ?? [];
    flow.error = rec.error ? new FlowError(rec.error.msg, rec.error.timestamp) : null;
    const r = rec.request;
    flow.request.scheme = r.scheme;
    flow.request.host = r.host;
    flow.request.port = r.port;
    flow.request.method = r.method;
    flow.request.path = r.path;
    flow.request.httpVersion = r.httpVersion;
    flow.request.headers = new Headers(r.headers);
    flow.request.body = r.body ? Buffer.from(r.body, "base64") : null;
    flow.request.bodyTruncated = r.bodyTruncated;
    flow.request.timestampStart = r.timestampStart;
    flow.request.timestampEnd = r.timestampEnd;
    if (rec.response) {
      const res = new CnResponse();
      res.statusCode = rec.response.statusCode;
      res.reason = rec.response.reason;
      res.httpVersion = rec.response.httpVersion;
      res.headers = new Headers(rec.response.headers);
      res.body = rec.response.body ? Buffer.from(rec.response.body, "base64") : null;
      res.bodyTruncated = rec.response.bodyTruncated;
      res.timestampStart = rec.response.timestampStart;
      res.timestampEnd = rec.response.timestampEnd;
      flow.response = res;
    }
    flow.websocketMessages = rec.websocketMessages.map((m) => ({
      fromClient: m.fromClient,
      type: m.type,
      content: Buffer.from(m.content, "base64"),
      timestamp: m.timestamp,
    }));
    return flow;
  }

  /** Serializable summary for the flow list (no bodies). */
  toSummary(): FlowSummary {
    return {
      id: this.id,
      type: this.type,
      method: this.request.method,
      scheme: this.request.scheme,
      host: this.request.prettyHost,
      path: this.request.path,
      url: this.request.url,
      statusCode: this.response?.statusCode ?? null,
      contentType: this.response?.contentType ?? "",
      reqSize: this.request.body?.length ?? 0,
      resSize: this.response?.body?.length ?? 0,
      duration: this.duration,
      timestamp: this.timestampCreated,
      marked: this.marked,
      mocked: this.mocked,
      intercepted: this.intercepted,
      error: this.error?.msg ?? null,
      appliedRules: this.appliedRules,
      wsMessages: this.type === "websocket" ? this.websocketMessages.length : 0,
    };
  }

  /** Full detail including decoded-as-base64 bodies. */
  toDetail(): FlowDetail {
    return {
      ...this.toSummary(),
      request: {
        httpVersion: this.request.httpVersion,
        headers: this.request.headers.entries(),
        body: this.request.body ? this.request.body.toString("base64") : null,
        bodyTruncated: this.request.bodyTruncated,
      },
      response: this.response
        ? {
            statusCode: this.response.statusCode,
            reason: this.response.reason,
            httpVersion: this.response.httpVersion,
            headers: this.response.headers.entries(),
            body: this.response.body ? this.response.body.toString("base64") : null,
            bodyTruncated: this.response.bodyTruncated,
          }
        : null,
      websocketMessages: this.websocketMessages.map((m) => ({
        fromClient: m.fromClient,
        type: m.type,
        content: m.content.toString("base64"),
        timestamp: m.timestamp,
      })),
      client: this.client,
      comment: this.comment,
    };
  }
}

export interface FlowRecord {
  id: string;
  type: FlowType;
  timestampCreated: number;
  marked: boolean;
  comment: string;
  mocked: boolean;
  appliedRules: string[];
  error: { msg: string; timestamp: number } | null;
  client: ClientInfo;
  request: {
    scheme: "http" | "https";
    host: string;
    port: number;
    method: string;
    path: string;
    httpVersion: string;
    headers: [string, string][];
    body: string | null;
    bodyTruncated: boolean;
    timestampStart: number;
    timestampEnd: number;
  };
  response: {
    statusCode: number;
    reason: string;
    httpVersion: string;
    headers: [string, string][];
    body: string | null;
    bodyTruncated: boolean;
    timestampStart: number;
    timestampEnd: number;
  } | null;
  websocketMessages: { fromClient: boolean; type: "text" | "binary"; content: string; timestamp: number }[];
}

export interface FlowSummary {
  id: string;
  type: FlowType;
  method: string;
  scheme: string;
  host: string;
  path: string;
  url: string;
  statusCode: number | null;
  contentType: string;
  reqSize: number;
  resSize: number;
  duration: number | null;
  timestamp: number;
  marked: boolean;
  mocked: boolean;
  intercepted: boolean;
  error: string | null;
  appliedRules: string[];
  wsMessages: number;
}

export interface FlowDetail extends FlowSummary {
  request: {
    httpVersion: string;
    headers: [string, string][];
    body: string | null;
    bodyTruncated: boolean;
  };
  response: {
    statusCode: number;
    reason: string;
    httpVersion: string;
    headers: [string, string][];
    body: string | null;
    bodyTruncated: boolean;
  } | null;
  websocketMessages: { fromClient: boolean; type: string; content: string; timestamp: number }[];
  client: ClientInfo;
  comment: string;
}

File: cnproxy-core/src/flow.rs
//! Flow — a single captured HTTP request/response pair.

use serde::{Deserialize, Serialize};

/// Unique flow identifier.
pub type FlowId = String;

/// Summary sent to the web inspector for the flow list.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FlowSummary {
    pub id: FlowId,
    pub method: String,
    pub url: String,
    pub host: String,
    pub path: String,
    pub status_code: Option<u16>,
    pub content_type: Option<String>,
    pub req_size: u64,
    pub res_size: u64,
    pub duration: Option<u64>, // ms
    pub flow_type: Option<String>,
    pub ws_messages: Option<u64>,
    pub mocked: bool,
    pub intercepted: bool,
    pub error: Option<String>,
    pub color: Option<String>,
    pub applied_rules: Vec<String>,
}

/// Full flow detail (headers, bodies, timings).
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Flow {
    pub id: FlowId,
    pub method: String,
    pub url: String,
    pub host: String,
    pub path: String,
    pub client: ClientInfo,
    pub request: RequestInfo,
    pub response: Option<ResponseInfo>,
    pub timing: Option<TimingInfo>,
    pub flow_type: String,
    pub ws_messages: Vec<WsMessage>,
    pub intercepted: bool,
    pub mocked: bool,
    pub error: Option<String>,
    pub color: Option<String>,
    pub applied_rules: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ClientInfo {
    pub address: String,
    pub port: u16,
    pub tls: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RequestInfo {
    pub headers: Vec<(String, String)>,
    pub body: Option<String>, // base64
    pub body_truncated: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ResponseInfo {
    pub status_code: u16,
    pub reason: String,
    pub headers: Vec<(String, String)>,
    pub body: Option<String>, // base64
    pub body_truncated: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TimingInfo {
    pub dns: Option<u64>,
    pub connect: Option<u64>,
    pub tls: Option<u64>,
    pub ttfb: Option<u64>,
    pub total: Option<u64>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct WsMessage {
    pub from_client: bool,
    pub content: String, // base64
    pub msg_type: String, // "text" | "binary"
    pub timestamp: u64,   // ms epoch
}

impl Flow {
    /// Convert to a lightweight summary for the flow list.
    pub fn to_summary(&self) -> FlowSummary {
        FlowSummary {
            id: self.id.clone(),
            method: self.method.clone(),
            url: self.url.clone(),
            host: self.host.clone(),
            path: self.path.clone(),
            status_code: self.response.as_ref().map(|r| r.status_code),
            content_type: self.response.as_ref().and_then(|r| {
                r.headers.iter().find(|(k, _)| k.eq_ignore_ascii_case("content-type")).map(|(_, v)| v.clone())
            }),
            req_size: self.request.body.as_ref().map(|b| b.len() as u64).unwrap_or(0),
            res_size: self.response.as_ref().and_then(|r| r.body.as_ref().map(|b| b.len() as u64)).unwrap_or(0),
            duration: self.timing.as_ref().and_then(|t| t.total),
            flow_type: if self.flow_type.is_empty() { None } else { Some(self.flow_type.clone()) },
            ws_messages: if self.ws_messages.is_empty() { None } else { Some(self.ws_messages.len() as u64) },
            mocked: self.mocked,
            intercepted: self.intercepted,
            error: self.error.clone(),
            color: self.color.clone(),
            applied_rules: self.applied_rules.clone(),
        }
    }
}